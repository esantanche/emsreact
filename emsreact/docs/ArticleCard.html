<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: ArticleCard</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: ArticleCard</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>ArticleCard<span class="signature">(classes, title, image, text_content, article_node_id)</span><span class="type-signature"></span></h2>
        
            <div class="class-description"><p>The card containing an article, or better, just the first part of it.</p>
<p>The article is rendered as pure text with no html.</p>
<p>It's also truncated and ellipses are added at the truncation point.</p>
<p>Props</p></div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="ArticleCard"><span class="type-signature"></span>new ArticleCard<span class="signature">(classes, title, image, text_content, article_node_id)</span><span class="type-signature"></span></h4>
    

    











    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>classes</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last"><p>Classes injected by withStyles function</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>title</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last"><p>Title of the article</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>image</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last"><p>Relative url of image as stored in Drupal file system</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>text_content</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last"><p>Full text of article including html tags</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>article_node_id</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            

            

            <td class="description last"><p>Node id of article to be used to retrieve
the article from the backend</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="components_library_cards_ArticleCard.js.html">components_library/cards/ArticleCard.js</a>, <a href="components_library_cards_ArticleCard.js.html#line74">line 74</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="produce_jsx_fragment_for_image_and_text_according_to_width"><span class="type-signature"></span>produce_jsx_fragment_for_image_and_text_according_to_width<span class="signature">(width, classes, image, text_content)</span><span class="type-signature"> &rarr; {jsx}</span></h4>
    

    



<div class="description">
    <p>This function produces the jsx fragment that displays the image and the text of the article.</p>
<p>It put the image above the text for small devices and the image on the left and the text on the right
for large devices.</p>
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>width</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last"><p>Material UI width (@see https://material-ui.com/layout/breakpoints/#withwidth-)</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>classes</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last"><p>Classes to be used to make the image responsive</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>image</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last"><p>Relative url of image as stored in Drupal file system</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>text_content</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last"><p>Full text of article including html tags</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="components_library_cards_ArticleCard.js.html">components_library/cards/ArticleCard.js</a>, <a href="components_library_cards_ArticleCard.js.html#line96">line 96</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Returns:</h5>

        
<div class="param-desc">
    <p>Jsx fragment showing the image and the text of the article</p>
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">jsx</span>


    </dd>
</dl>

    





        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="AlignedGridPane.html">AlignedGridPane</a></li><li><a href="App.html">App</a></li><li><a href="ArticleCard.html">ArticleCard</a></li><li><a href="ArticleScreen.html">ArticleScreen</a></li><li><a href="ArticleService.html">ArticleService</a></li><li><a href="BodyPane.html">BodyPane</a></li><li><a href="BottomGridPane.html">BottomGridPane</a></li><li><a href="ColouredCard.html">ColouredCard</a></li><li><a href="FooterScreen.html">FooterScreen</a></li><li><a href="FrontGridPane.html">FrontGridPane</a></li><li><a href="HeaderMenu.html">HeaderMenu</a></li><li><a href="HeaderScreen.html">HeaderScreen</a></li><li><a href="HomeButtonPane.html">HomeButtonPane</a></li><li><a href="HomeButtonText.html">HomeButtonText</a></li><li><a href="HomeScreen.html">HomeScreen</a></li><li><a href="InnerGridPane.html">InnerGridPane</a></li><li><a href="ListOfArticlesScreen.html">ListOfArticlesScreen</a></li><li><a href="MasterAppBar.html">MasterAppBar</a></li><li><a href="MasterAvatar.html">MasterAvatar</a></li><li><a href="MenuButton.html">MenuButton</a></li><li><a href="MessageDialog.html">MessageDialog</a></li><li><a href="PickMyBrainDialog.html">PickMyBrainDialog</a></li><li><a href="PickMyBrainMessageService.html">PickMyBrainMessageService</a></li><li><a href="PickMyBrainScreen.html">PickMyBrainScreen</a></li><li><a href="SeparatorPane.html">SeparatorPane</a></li><li><a href="SocialCard.html">SocialCard</a></li><li><a href="SocialScreen.html">SocialScreen</a></li><li><a href="TopicScreen.html">TopicScreen</a></li><li><a href="TopicService.html">TopicService</a></li></ul><h3>Global</h3><ul><li><a href="global.html#theme">theme</a></li><li><a href="global.html#title_to_slug">title_to_slug</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Wed Aug 29 2018 11:52:02 GMT+0000 (UTC)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>