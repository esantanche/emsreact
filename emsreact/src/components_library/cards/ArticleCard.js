// Just a test to refactor
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Material UI functions that make classes and width available
import { withStyles } from "@material-ui/core/styles/index";
import withWidth from '@material-ui/core/withWidth';

// Material UI components
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardMedia from '@material-ui/core/CardMedia';
import Grid from '@material-ui/core/Grid';
import CardActions from '@material-ui/core/CardActions';
import IconButton from '@material-ui/core/IconButton';

// Material UI icons
import MoreHoriz from '@material-ui/icons/MoreHoriz';

// App configuration
import { APP_CONFIGURATION } from '../../appConfiguration';

// Library used to strip html tags from text
import striptags from 'striptags';

// Used to cut text and add '...'
import Dotdotdot from 'react-dotdotdot';

// From component library
import SeparatorPane from '../panes/SeparatorPane';

// Helper function that takes a title and produces a slug
// which is how we put the title in the url
import { title_to_slug } from '../../helpers/title_to_slug';

const styles = {
    card_header: {
        height: "6em",
        display: "flex",
        alignItems: "center",
        paddingLeft: "24px",
        paddingRight: "24px"
    },
    container: {
        display: "flex",
        flexDirection: "row",
        alignItems: "stretch"
    },
    image_responsive_size: {
        height: "0",
        paddingTop: '100%'
    }
};

class ArticleCard extends Component {

    static propTypes = {
        classes: PropTypes.object.isRequired,
        title: PropTypes.string,
        image: PropTypes.string,
        text_content: PropTypes.string,
        article_node_id: PropTypes.number
    };

    produce_jsx_fragment_for_image_and_text_according_to_width(width, classes, image, text_content) {

        const image_fixed_size_according_to_width = {
            sm: "150px",
            md: "150px",
            lg: "180px",
            xl: "200px"
        };

        // We are going to remove any html tag from the text of the article
        // because we want no html tags when showing the text in the article
        // card
        const stripped_text_content = striptags(text_content);

        // When fetching images' url from the Drupal backend, we get relative paths
        // and we have to prepend the backend root url to access the images
        // The module Pathologic in Drupal makes urls absolutes, but not in this case

        if (width === 'xs') {

            return (
                <React.Fragment>

                    <Grid item xs={12}>

                        <CardMedia
                            className={classes.image_responsive_size}
                            image={APP_CONFIGURATION.backendUrl + image}
                            title=""
                        />

                    </Grid>

                    <SeparatorPane/>

                    <Grid item xs={12}>

                        <Typography variant="body2" component="div">

                            <Dotdotdot clamp={5}>

                                {stripped_text_content}

                            </Dotdotdot>

                        </Typography>

                    </Grid>

                </React.Fragment>
            );

        } else {

            // In this case we want the text to be as high as the image.
            // The dotdotdot component will cut the text and add '...'
            // at the end when the text is as high as the image

            return (
                <React.Fragment>

                    <CardMedia
                        style={{ height: image_fixed_size_according_to_width[width],
                                 width: image_fixed_size_according_to_width[width] }}
                        // className={classes.image_fixed_size}
                        image={APP_CONFIGURATION.backendUrl + image}
                        title=""
                    />

                    <Typography variant="body2" component="div"
                                style={{ flex: "1", height: image_fixed_size_according_to_width[width], marginLeft: "10px" }}>

                        <Dotdotdot clamp={image_fixed_size_according_to_width[width]}
                                   style={{ height: image_fixed_size_according_to_width[width] }}>

                            {stripped_text_content}

                        </Dotdotdot>

                    </Typography>

                </React.Fragment>
            );

        }

    }

    render() {

        const title_variant_according_to_width = {
            xs: "title",
            sm: "title",
            md: "subheading",
            lg: "title",
            xl: "title"
        };

        const { classes, title, width, image, text_content, article_node_id } = this.props;

        const jsx_fragment_for_image_and_text_according_to_width =
            this.produce_jsx_fragment_for_image_and_text_according_to_width(width, classes, image, text_content);

        // FIXME link to article to fix

        const slug_for_link_to_article = title_to_slug(title);

        return (

            <Card>

                {/*Used a div because the component CardHeader uses a headline variant of Typography while I    */}
                {/*need the possibility to change font size*/}
                {/*style={{ height: "6em", display: "flex", alignItems: "center", paddingLeft: "24px", paddingRight: "24px" }}*/}

                <div className={classes.card_header}>

                    <Typography variant={title_variant_according_to_width[width]}>
                        {title}
                    </Typography>

                </div>

                <CardContent>

                    <Grid container spacing={0} className={classes.container}>

                        {jsx_fragment_for_image_and_text_according_to_width}

                    </Grid>

                </CardContent>

                <CardActions>
                    <IconButton style={{ marginLeft: "auto" }} href={"/article/" + article_node_id + "/" + slug_for_link_to_article}>
                        <MoreHoriz/>
                    </IconButton>
                </CardActions>

            </Card>

        )

    }

}

export default withWidth()(withStyles(styles)(ArticleCard));
