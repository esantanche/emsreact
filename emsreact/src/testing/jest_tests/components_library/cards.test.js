// FIXME docs

import ReactDOM from "react-dom";
import ArticleCard from "../../../components_library/cards/ArticleCard";
import React from "react";
import renderer from 'react-test-renderer';

// Here we test components in components_library/cards

// Testing ArticleCard

const article_card_to_be_tested = ( <ArticleCard title="Article card title"
                                                 image="http://backend.emanuelesantanche.com/sites/default/files/2018-07/css-in-js_0.png"
                                                 text_content="Article card text content"
                                                 article_node_id={22}/> );

it('ArticleCard renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(article_card_to_be_tested, div);
    ReactDOM.unmountComponentAtNode(div);
});

// Getting rid of react-dotdotdot library that causes an error
jest.mock('react-dotdotdot', () => 'div');

it('ArticleCard renders correctly', () => {
    const tree = renderer
        .create(article_card_to_be_tested)
        .toJSON();
    expect(tree).toMatchSnapshot();
});