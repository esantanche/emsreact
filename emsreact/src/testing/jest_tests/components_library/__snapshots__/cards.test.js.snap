// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard renders correctly 1`] = `
<div
  className="MuiPaper-root-186 MuiPaper-elevation1-189 MuiPaper-rounded-187 MuiCard-root-185"
>
  <div
    className="ArticleCard-card_header-181"
  >
    <h3
      className="MuiTypography-root-213 MuiTypography-subheading-220"
    >
      Article card title
    </h3>
  </div>
  <div
    className="MuiCardContent-root-237"
  >
    <div
      className="MuiGrid-container-238 ArticleCard-container-182"
    >
      <div
        className="MuiCardMedia-root-334"
        src={undefined}
        style={
          Object {
            "backgroundImage": "url(\\"http://backend.emanuelesantanche.comhttp://backend.emanuelesantanche.com/sites/default/files/2018-07/css-in-js_0.png\\")",
            "height": "150px",
            "width": "150px",
          }
        }
        title=""
      />
      <div
        className="MuiTypography-root-213 MuiTypography-body2-221"
        style={
          Object {
            "flex": "1",
            "height": "150px",
            "marginLeft": "10px",
          }
        }
      >
        <div
          clamp="150px"
          style={
            Object {
              "height": "150px",
            }
          }
        >
          Article card text content
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiCardActions-root-336"
  >
    <a
      className="MuiButtonBase-root-344 MuiIconButton-root-338 ArticleCard-margin_for_more_button-184 MuiCardActions-action-337"
      href="/article/22/article-card-title"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="MuiIconButton-label-343"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-347"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g>
            <path
              d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </g>
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-354"
      />
    </a>
  </div>
</div>
`;
